File: adapters/java/src/test/java/adapter/DcborTest.java
package adapter;

import static org.junit.jupiter.api.Assertions.*;

import java.math.BigInteger;
import java.util.List;
import org.junit.jupiter.api.Test;

class DcborTest {
    private static LogicalValue.Flt f(String value) {
        return new LogicalValue.Flt("auto", value);
    }

    @Test
    void d5FloatWholeBecomesInt() {
        assertArrayEquals(Util.hexDecode("02"), Dcbor.encode(f("2.0")));
    }

    @Test
    void d5FloatFractionStaysFloat() {
        assertArrayEquals(Util.hexDecode("f94100"), Dcbor.encode(f("2.5")));
    }

    @Test
    void d6NanCanonicalPayload() {
        assertArrayEquals(Util.hexDecode("f97e00"), Dcbor.encode(f("NaN")));
    }

    @Test
    void d7ZeroAsInt() {
        assertArrayEquals(Util.hexDecode("00"), Dcbor.encode(new LogicalValue.Int("0")));
    }

    @Test
    void d7ZeroAsPosFloatUnifiesWithInt() {
        assertArrayEquals(Util.hexDecode("00"), Dcbor.encode(f("0.0")));
    }

    @Test
    void d7ZeroAsNegFloatUnifiesWithInt() {
        // Opposite of rfc8949's P7: -0.0 must also collapse to plain int 0.
        assertArrayEquals(Util.hexDecode("00"), Dcbor.encode(f("-0.0")));
    }

    @Test
    void d8NfcCombiningAccentNormalizes() {
        LogicalValue v = new LogicalValue.Text("café");
        assertArrayEquals(Util.hexDecode("65636166c3a9"), Dcbor.encode(v));
    }

    @Test
    void bignumBelowNativeRangeIsRejected() {
        LogicalValue value = new LogicalValue.Bignum("positive", "5");
        assertThrows(Rfc8949.EncodeException.class, () -> Dcbor.encode(value));
    }

    @Test
    void bignumAt2Pow64UsesTag2() {
        LogicalValue value = new LogicalValue.Bignum("positive", "18446744073709551616");
        assertArrayEquals(Util.hexDecode("c249010000000000000000"), Dcbor.encode(value));
    }

    @Test
    void tagWrapsInnerValue() {
        LogicalValue value = new LogicalValue.Tag(BigInteger.valueOf(100), new LogicalValue.Int("5"));
        assertArrayEquals(new byte[] { (byte) 0xd8, 0x64, 0x05 }, Dcbor.encode(value));
    }

    @Test
    void reservedBignumTagViaGenericTagArmIsARawPassthrough() {
        LogicalValue value = new LogicalValue.Tag(BigInteger.valueOf(2), new LogicalValue.Bytes("01"));
        assertArrayEquals(new byte[] { (byte) 0xc2, 0x41, 0x01 }, Dcbor.encode(value));
    }

    @Test
    void explicitWidthIsIgnoredWholeFloatStillReduces() {
        LogicalValue v = new LogicalValue.Flt("f64", "2.0");
        assertArrayEquals(Util.hexDecode("02"), Dcbor.encode(v));
    }

    @Test
    void explicitWidthIsIgnoredNanStillCanonical() {
        LogicalValue v = new LogicalValue.Flt("f32", "NaN");
        assertArrayEquals(Util.hexDecode("f97e00"), Dcbor.encode(v));
    }

    @Test
    void tagAtU64MaxDoesNotTruncate() {
        LogicalValue value = new LogicalValue.Tag(new BigInteger("18446744073709551615"), new LogicalValue.Int("5"));
        assertArrayEquals(Util.hexDecode("dbffffffffffffffff05"), Dcbor.encode(value));
    }

    @Test
    void mapKeyCollisionDedupesLastWriteWins() {
        // D7: 0, 0.0, and -0.0 all canonicalize to encoded key 0x00, so they
        // collapse to one entry, keeping only the last value written (3).
        LogicalValue value = new LogicalValue.Map(List.of(
            java.util.Map.entry(new LogicalValue.Int("0"), new LogicalValue.Int("1")),
            java.util.Map.entry(f("0.0"), new LogicalValue.Int("2")),
            java.util.Map.entry(f("-0.0"), new LogicalValue.Int("3"))));
        assertArrayEquals(Util.hexDecode("a10003"), Dcbor.encode(value));
    }

    @Test
    void mapKeyCollisionViaNfcDedupesLastWriteWins() {
        // D8: "café" (precomposed) and "café" (NFC-decomposed) both normalize
        // to the identical encoded key bytes, so they must also collapse to a
        // single entry, last write wins.
        String decomposed = "café"; // "e" + combining acute accent
        String precomposed = "café"; // single precomposed char
        LogicalValue value = new LogicalValue.Map(List.of(
            java.util.Map.entry(new LogicalValue.Text(decomposed), new LogicalValue.Int("1")),
            java.util.Map.entry(new LogicalValue.Text(precomposed), new LogicalValue.Int("2"))));
        assertArrayEquals(Util.hexDecode("a165636166c3a902"), Dcbor.encode(value));
    }

    @Test
    void arrayAndMapRoundTrip() {
        LogicalValue.Arr arr = new LogicalValue.Arr(List.of(new LogicalValue.Int("1"), new LogicalValue.Int("2")));
        assertArrayEquals(new byte[] { (byte) 0x82, 0x01, 0x02 }, Dcbor.encode(arr));

        LogicalValue map = new LogicalValue.Map(
            List.of(java.util.Map.entry(new LogicalValue.Text("a"), new LogicalValue.Int("1"))));
        assertArrayEquals(Util.hexDecode("a1616101"), Dcbor.encode(map));
    }
}

File: adapters/kotlin/src/test/kotlin/DcborTest.kt
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class DcborTest {
    private fun float(value: String) = LogicalValue.Float("auto", value)

    @Test
    fun d5FloatWholeBecomesInt() {
        assertEquals(hexDecode("02").toList(), encodeDcbor(float("2.0")).toList())
    }

    @Test
    fun d5FloatFractionStaysFloat() {
        assertEquals(hexDecode("f94100").toList(), encodeDcbor(float("2.5")).toList())
    }

    @Test
    fun d6NanCanonicalPayload() {
        assertEquals(hexDecode("f97e00").toList(), encodeDcbor(float("NaN")).toList())
    }

    @Test
    fun d7ZeroAsInt() {
        assertEquals(hexDecode("00").toList(), encodeDcbor(LogicalValue.Int("0")).toList())
    }

    @Test
    fun d7ZeroAsPosFloatUnifiesWithInt() {
        assertEquals(hexDecode("00").toList(), encodeDcbor(float("0.0")).toList())
    }

    @Test
    fun d7ZeroAsNegFloatUnifiesWithInt() {
        // Opposite of rfc8949's P7: -0.0 must also collapse to plain int 0.
        assertEquals(hexDecode("00").toList(), encodeDcbor(float("-0.0")).toList())
    }

    @Test
    fun d8NfcCombiningAccentNormalizes() {
        val v = LogicalValue.Text("café")
        assertEquals(hexDecode("65636166c3a9").toList(), encodeDcbor(v).toList())
    }

    @Test
    fun bignumBelowNativeRangeIsRejected() {
        val value = LogicalValue.Bignum("positive", "5")
        assertFailsWith<EncodeException> { encodeDcbor(value) }
    }

    @Test
    fun bignumAt2Pow64UsesTag2() {
        val value = LogicalValue.Bignum("positive", "18446744073709551616")
        assertEquals(hexDecode("c249010000000000000000").toList(), encodeDcbor(value).toList())
    }

    @Test
    fun tagWrapsInnerValue() {
        val value = LogicalValue.Tag(100uL, LogicalValue.Int("5"))
        assertEquals(listOf(0xd8, 0x64, 0x05).map { it.toByte() }, encodeDcbor(value).toList())
    }

    @Test
    fun reservedBignumTagViaGenericTagArmIsARawPassthrough() {
        val value = LogicalValue.Tag(2uL, LogicalValue.Bytes("01"))
        assertEquals(listOf(0xc2, 0x41, 0x01).map { it.toByte() }, encodeDcbor(value).toList())
    }

    @Test
    fun explicitWidthIsIgnoredWholeFloatStillReduces() {
        val v = LogicalValue.Float("f64", "2.0")
        assertEquals(hexDecode("02").toList(), encodeDcbor(v).toList())
    }

    @Test
    fun explicitWidthIsIgnoredNanStillCanonical() {
        val v = LogicalValue.Float("f32", "NaN")
        assertEquals(hexDecode("f97e00").toList(), encodeDcbor(v).toList())
    }

    @Test
    fun tagAtULongMaxDoesNotTruncate() {
        val value = LogicalValue.Tag(ULong.MAX_VALUE, LogicalValue.Int("5"))
        assertEquals(hexDecode("dbffffffffffffffff05").toList(), encodeDcbor(value).toList())
    }

    @Test
    fun mapKeyCollisionDedupesLastWriteWins() {
        // D7: 0, 0.0, and -0.0 all canonicalize to encoded key 0x00, so the
        // reference dcbor crate's Map::insert collapses these three logical
        // entries to one, keeping only the last value written (3). Mirrors
        // adapters/rust/src/dcbor.rs's Map::insert semantics.
        val value = LogicalValue.Map(
            listOf(
                LogicalValue.Int("0") to LogicalValue.Int("1"),
                float("0.0") to LogicalValue.Int("2"),
                float("-0.0") to LogicalValue.Int("3"),
            )
        )
        assertEquals(hexDecode("a10003").toList(), encodeDcbor(value).toList())
    }

    @Test
    fun mapKeyCollisionViaNfcDedupesLastWriteWins() {
        // D8: "café" (precomposed) and "café" (NFC-decomposed) both
        // normalize to the identical encoded key bytes, so they must also
        // collapse to a single entry, last write wins.
        val decomposed = "café" // "e" + combining acute accent (U+0301)
        val precomposed = "café" // single precomposed char (U+00E9)
        val value = LogicalValue.Map(
            listOf(
                LogicalValue.Text(decomposed) to LogicalValue.Int("1"),
                LogicalValue.Text(precomposed) to LogicalValue.Int("2"),
            )
        )
        assertEquals(hexDecode("a165636166c3a902").toList(), encodeDcbor(value).toList())
    }

    @Test
    fun arrayAndMapRoundTrip() {
        val arr = LogicalValue.Array(listOf(LogicalValue.Int("1"), LogicalValue.Int("2")))
        assertEquals(listOf(0x82, 0x01, 0x02).map { it.toByte() }, encodeDcbor(arr).toList())

        val map = LogicalValue.Map(listOf(LogicalValue.Text("a") to LogicalValue.Int("1")))
        assertEquals(hexDecode("a1616101").toList(), encodeDcbor(map).toList())
    }
}
